### PR TITLE
8292077: G1 nmethod entry barriers don't keep oops alive

### PR DESCRIPTION
**Pull request by fisk**
The intention is that when G1 uses nmethod entry barriers (currently with --enable-preview to support Loom, but soon also with https://bugs.openjdk.org/browse/JDK-8290025), it keeps nmethod oops alive the first time an nmethod becomes on-stack during concurrent marking. The current code fails to do that, even though it intended to. That should be fixed.

There was code from loom to keep the oops alive by performing a phantom load. However, there is a quirk with the access API; if you don't assign the result of the load to something, the load won't be performed, and hence the oop won't be kept alive. I changed the code to use CollectedHeap::keep_alive explicitly instead of relying on side effects of a dummy load.
I also found that we assume G1 concurrent marking isn't aborted more than once. But that can happen. I made the assumption more conservative - surely if there are INT_MAX - 1 interrupted concurrent GCs in a row... something is very very wrong.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292077](https://bugs.openjdk.org/browse/JDK-8292077): G1 nmethod entry barriers don't keep oops alive

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [7f2f347e](https://git.openjdk.org/jdk/pull/9817/files/7f2f347e0949eb2d23ce4473f7676f43a5132700)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9817/head:pull/9817` \
`$ git checkout pull/9817`

Update a local copy of the PR: \
`$ git checkout pull/9817` \
`$ git pull https://git.openjdk.org/jdk pull/9817/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9817`

View PR using the GUI difftool: \
`$ git pr show -t 9817`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9817.diff">https://git.openjdk.org/jdk/pull/9817.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp [hunk 1]
```diff
@@ -38,11 +40,19 @@
 #include "runtime/threads.hpp"
 #include "utilities/debug.hpp"
 
-class LoadPhantomOopClosure : public OopClosure {
+class OopKeepaliveClosure : public OopClosure {
+private:
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/barrierSetNMethod.cpp:44]:
```suggestion
```
This line can be dropped.

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp [hunk 2]
```diff
@@ -105,8 +114,14 @@ class BarrierSetNMethodArmClosure : public ThreadClosure {
...
   // disarm value for all threads, to reflect the new GC phase.
+  // We wrap around at INT_MAX. That means that we assume nmethods won't have ABA
+  // problems in their nmethod disarm values after INT_MAX - 1 GCs. Every time a GC
+  // completes, ABA problems are removed, but if a concurrent GC is started and then
+  // aborted N times, that is when there could be ABA problems. If there are anything
+  // close to INT_MAX - 1 GCs starting without being able to finish, something is
+  // seriously wrong.
   ++_current_phase;
-  if (_current_phase == 4) {
+  if (_current_phase == INT_MAX) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/barrierSetNMethod.cpp:124]: I would really prefer to deal with this in a separate issue. It does not have anything to do with keeping oops alive in the nmethod barrier.

**Review by tschatzl — CHANGES_REQUESTED** [review]

**Comment by fisk** [comment]
Thanks for the review, @tschatzl . I applied your requested changes.

**Review by tschatzl — APPROVED** [review]
Ship it

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp [hunk 3]
```diff
@@ -38,11 +40,18 @@
...
+    : _heap(Universe::heap()) {}
+
   virtual void do_oop(oop* p) {
-    NativeAccess<ON_PHANTOM_OOP_REF>::oop_load(p);
+    oop obj = NativeAccess<ON_PHANTOM_OOP_REF | AS_NO_KEEPALIVE>::oop_load(p);
+    _heap->keep_alive(obj);
   }
+
   virtual void do_oop(narrowOop* p) { ShouldNotReachHere(); }
 };
```
Comment (stefank) [inline src/hotspot/share/gc/shared/barrierSetNMethod.cpp:56]: * Keepalive should probably be KeepAlive. (I realize that AS_NO_KEEPALIVE doesn't split the two words, but that one is the outlier in the HotSpot code base)
* Seems unnecessary to cache the heap pointer
* Could this class be moved into the BarrierSetNMethod scope. Or, alternatively give it a more descriptive name stating that this is for nmethods and/or native phantom loads?
 I'm thinking about future name collisions and ODR.
* Could you add a comment stating why you do this dance with AS_NO_KEEPALIVE + keep_alive(oop)

**Review by stefank — CHANGES_REQUESTED** [review]

**Comment by fisk** [comment]
Thanks for the review, @tschatzl

**Comment by fisk** [comment]
Thanks for the review, @stefank. I applied your requested changes.

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp [hunk 4]
```diff
@@ -67,10 +61,26 @@ bool BarrierSetNMethod::supports_entry_barrier(nmethod* nm) {
 }
 
 bool BarrierSetNMethod::nmethod_entry_barrier(nmethod* nm) {
+  class OopKeepAliveClosure : public OopClosure {
+  public:
+    OopKeepAliveClosure() {}
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/barrierSetNMethod.cpp:66]: The constructor could be removed as well now. Still looks good.

**Review by tschatzl — APPROVED** [review]
Still good.

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp (hunk 4 above)
Comment (fisk) [reply src/hotspot/share/gc/shared/barrierSetNMethod.cpp]: Fixed. Thanks Thomas.

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp [hunk 5]
```diff
@@ -67,10 +61,24 @@ bool BarrierSetNMethod::supports_entry_barrier(nmethod* nm) {
 }
 
 bool BarrierSetNMethod::nmethod_entry_barrier(nmethod* nm) {
+  class OopKeepAliveClosure : public OopClosure {
+  public:
+    virtual void do_oop(oop* p) {
+      // Loads on nmethod oops are phantom strength. The intend of the load
```
Comment (stefank) [inline src/hotspot/share/gc/shared/barrierSetNMethod.cpp:67]: intend => intent?

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp [hunk 6]
```diff
@@ -67,10 +61,24 @@ bool BarrierSetNMethod::supports_entry_barrier(nmethod* nm) {
...
+      // is to just read the oop, and then explicitly keep it alive w.r.t.
+      // concurrent marking. Using the keep alive side effects of a normal
+      // phantom load is less explicit, and doesn't actually do anything
+      // unless the returned value is used as an oop.
```
Comment (stefank) [inline src/hotspot/share/gc/shared/barrierSetNMethod.cpp:71]: I think this is bit vague. Would it be OK if we explicitly spelled out the problem with the LoadOopProxy? Maybe something like this (feel free to tweak the actual text):
```
// Loads on nmethod oops are phantom strength.
//
// Note: that we could have used NativeAccess<ON_PHANTOM_OOP_REF>::oop_load(p),
// but that would have *required* us to convert the returned LoadOopProxy to an oop,
// or else keep alive load barrier will never be called. It's the LoadOopProxy-to-oop
// conversion that performs the load barriers. This is too subtle, so we instead
// perform an explicit keep alive call.
```

**Review by stefank — APPROVED** [review]
Thanks. Would you mind a few more tweaks?

File: src/hotspot/share/gc/shared/barrierSetNMethod.cpp (hunk 6 above)
Comment (fisk) [reply src/hotspot/share/gc/shared/barrierSetNMethod.cpp]: I used your comment but removed the colon in "Note: that". Thanks for the review!

**Review by stefank — APPROVED** [review]

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by fisk** [comment]
Thanks for the reviews @iwanowww and @stefank!

**Pull request closed without merging**